File: frontend/src/api/dataQualityGenerate.ts
/**
 * 数据质量规则自动生成 — API client。
 * 对应后端 Task 4/5/6 端点。
 */

import { httpClient } from "./client";
import type {
  GeneratePreviewResponse,
  GenerateConfirmResponse,
  LlmModelOption,
  ParseDescriptionsResponse,
  RuleSuggestion,
} from "../types/dataQualityGenerate";

const BASE = "/data-quality/rules/generate";
const MODELS_BASE = "/models";

// ---------------------------------------------------------------------------
// Task 4: 预览规则建议
// ---------------------------------------------------------------------------

/**
 * 预览：给定本体类 + 数据源，调用 GeneratorService.preview，
 * 返回候选规则建议列表及被屏蔽的属性。
 */
export async function previewRules(
  classId: number,
  datasourceId: number,
): Promise<GeneratePreviewResponse> {
  const res = await httpClient.post<GeneratePreviewResponse>(`${BASE}/preview`, {
    classId,
    datasourceId,
  });
  return res.data;
}

// ---------------------------------------------------------------------------
// Task 5: 确认并创建规则
// ---------------------------------------------------------------------------

/**
 * 确认：批量写入用户采纳的规则建议，返回实际创建成功的规则列表
 * 及因重复而跳过的 ruleCode 列表。
 */
export async function confirmRules(
  datasourceId: number,
  rules: RuleSuggestion[],
): Promise<GenerateConfirmResponse> {
  const res = await httpClient.post<GenerateConfirmResponse>(`${BASE}/confirm`, {
    datasourceId,
    rules,
  });
  return res.data;
}

// ---------------------------------------------------------------------------
// Task 6: LLM 从属性描述中提取候选约束
// ---------------------------------------------------------------------------

/**
 * 列出当前可用的 LLM 模型配置（供向导下拉选择）。
 * 调用 GET /api/v1/models?activeOnly=true。
 *
 * 关键：query 参数名是 camelCase（与 FastAPI kwarg 名一致），
 * 写 snake_case 会被静默忽略，永远拿不到 activeOnly 过滤后的列表。
 */
export async function listLlmModels(): Promise<LlmModelOption[]> {
  const res = await httpClient.get<LlmModelOption[]>(MODELS_BASE, {
    params: { activeOnly: true },
  });
  return res.data.map((m) => ({
    id: m.id,
    modelName: m.modelName,
    provider: m.provider,
  }));
}

/**
 * 解析本体类的属性描述，LLM 推断候选约束（allowed_values / not_null）。
 * 返回完整 envelope（含 persistedPropertyIds），前端用它初始化 LlmPanel.adoptedIds。
 * `modelId` 可选：传入则路由层走对应 ModelConfig 创建 client，否则走默认 env 路径。
 */
export async function parseDescriptions(
  classId: number,
  modelId?: number | null,
): Promise<ParseDescriptionsResponse> {
  const res = await httpClient.post<ParseDescriptionsResponse>(
    `${BASE}/parse-descriptions`,
    { classId, modelId: modelId ?? null },
  );
  return res.data;
}

/**
 * 采纳 LLM 推荐的约束，按 kind 派发写入 ontology_property 表（feat-ontology-property-constraints）。
 * - allowed_values：传 allowedValues
 * - not_null      ：无需额外字段
 * - range         ：传 minValue + maxValue
 * - pattern       ：传 regexPattern
 *
 * 后端 schema 默认 kind=allowed_values 以兼容旧 client。
 */
export type SuggestionKind =
  | "allowed_values"
  | "not_null"
  | "range"
  | "pattern";

export interface ApplySuggestionPayload {
  kind: SuggestionKind;
  allowedValues?: string[];
  minValue?: string;
  maxValue?: string;
  regexPattern?: string;
}

export async function applySuggestion(
  propertyId: number,
  payload: ApplySuggestionPayload | string[],
): Promise<void> {
  // 兼容旧调用：传 string[] 时当作 allowed_values。
  const body: Record<string, unknown> = { propertyId };
  if (Array.isArray(payload)) {
    body.kind = "allowed_values";
    body.allowedValues = payload;
  } else {
    body.kind = payload.kind;
    if (payload.kind === "allowed_values") body.allowedValues = payload.allowedValues ?? [];
    if (payload.kind === "range") {
      body.minValue = payload.minValue;
      body.maxValue = payload.maxValue;
    }
    if (payload.kind === "pattern") body.regexPattern = payload.regexPattern;
  }
  await httpClient.post(`${BASE}/apply-suggestion`, body);
}

File: backend/app/api/v1/data_quality_generate.py
"""数据质量规则自动生成 API（dq-rule-auto-generation Task 4-6）。

POST /api/v1/data-quality/rules/generate/preview   预览规则建议
POST /api/v1/data-quality/rules/generate/confirm   确认并写入（Task 5）
POST /api/v1/data-quality/rules/generate/parse-descriptions  LLM 解析属性描述（Task 6）
POST /api/v1/data-quality/rules/generate/apply-suggestion  采纳 LLM 建议写入元数据（Task 6）
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import CurrentUser, getCurrentUser, getDb
from app.domain.schemas import (
    ApplySuggestionRequest,
    ApplySuggestionResponse,
    GenerateConfirmRequest,
    GenerateConfirmResponse,
    GeneratePreviewRequest,
    GeneratePreviewResponse,
    ParseDescriptionsRequest,
    ParseDescriptionsResponse,
)
from app.infrastructure.llm.factory import createClient
from app.services.data_quality_rule_generate_service import (
    DataQualityRuleGenerateService,
)
from app.services.data_quality_rule_llm_service import parsePropertyDescriptions

router = APIRouter(prefix="", tags=["data-quality-generate"])


def _getDefaultLlmClient() -> Any:
    """获取默认 LLM 客户端（支持测试 monkeypatch）。

    优先用 OPENAI_API_KEY 环境变量；若未配置则返回 None
   （路由层捕获 None → 503 LLMUnavailableError）。
    """
    return createClient(None)


@router.post("/preview", response_model=GeneratePreviewResponse)
async def previewRules(
    payload: GeneratePreviewRequest,
    user: CurrentUser = Depends(getCurrentUser),  # noqa: B008
    session: AsyncSession = Depends(getDb),  # noqa: B008
) -> GeneratePreviewResponse:
    """预览指定本体类的数据质量规则建议。

    返回该类可推导的全部规则建议（NEW/EXISTS 状态），以及因 schema 映射问题被阻断的属性列表。
    """
    return await DataQualityRuleGenerateService().preview(
        session, classId=payload.class_id, datasourceId=payload.datasource_id,
    )


@router.post("/confirm", response_model=GenerateConfirmResponse)
async def confirmRules(
    payload: GenerateConfirmRequest,
    user: CurrentUser = Depends(getCurrentUser),  # noqa: B008
    session: AsyncSession = Depends(getDb),  # noqa: B008
) -> GenerateConfirmResponse:
    """批量确认并写入自动生成的数据质量规则。

    幂等写入：rule_code 已存在 → 记入 skippedCodes；并发撞唯一约束同理。
    每条创建成功的规则会写入一条 audit_outbox 审计事件。
    """
    return await DataQualityRuleGenerateService().confirm(
        session, payload=payload, actor=user,
    )


@router.post("/parse-descriptions", response_model=ParseDescriptionsResponse)
async def parseDescriptions(
    payload: ParseDescriptionsRequest,
    user: CurrentUser = Depends(getCurrentUser),  # noqa: B008
    session: AsyncSession = Depends(getDb),  # noqa: B008
) -> ParseDescriptionsResponse:
    """LLM 解析本体类所有属性的 description 字段，推导候选约束建议。

    advisory 只读，不写 outbox。LLM 失败 → 503。
    """
    # 先用 service 做 classId 存在性校验（404）
    from app.domain.models import OntologyClass
    cls = await session.get(OntologyClass, payload.class_id)
    if cls is None:
        from app.domain.exceptions import NotFoundError
        from app.services.messages_zh import MSG_DQ_GEN_CLASS_NOT_FOUND
        raise NotFoundError(MSG_DQ_GEN_CLASS_NOT_FOUND.format(id=payload.class_id))

    # model_id 有值 → 按 LlmConfig 加载指定 provider；否则走默认 env（向后兼容）。
    # createClient 在 key 缺失时返回 None（不再让 OpenAI SDK 构造期 raise → 500），
    # 统一走「未配置 LLM → 503 LLMUnavailableError」语义。
    if payload.model_id:
        from app.services.model_config_service import ModelConfigService
        config = await ModelConfigService().get(session, payload.model_id)
        llm_client = createClient(config)
    else:
        llm_client = _getDefaultLlmClient()
    if llm_client is None:
        from app.domain.exceptions import LLMUnavailableError
        raise LLMUnavailableError("未配置 LLM，无法进行 AI 辅助分析")
    return await parsePropertyDescriptions(
        session, payload=payload, llm_client=llm_client, actor=user,
    )


@router.post("/apply-suggestion", response_model=ApplySuggestionResponse)
async def applySuggestion(
    payload: ApplySuggestionRequest,
    user: CurrentUser = Depends(getCurrentUser),  # noqa: B008
    session: AsyncSession = Depends(getDb),  # noqa: B008
) -> ApplySuggestionResponse:
    """采纳 LLM 推荐的 allowed_values，写入 ontology_property 并记录 outbox 审计。

    仅值域型约束写入元数据；业务必填类建议不写回。
    值不允许含单引号（SQL 注入防护）。
    """
    return await DataQualityRuleGenerateService().applySuggestion(
        session, payload=payload, actor=user,
    )
